Notification model

mattegan111 opened the pull request.

mattegan111 committed:
Add Notification Model

mattegan111 committed:
Add Notification CRUD routes

nathanrobb commented on server/models/Notification.js:
```diff
@@ -0,0 +1,31 @@
+		required: true,
+	},
+
+	notificationcontext: {
```
Why did you go with the name `context` here instead of something like `id`? Any reason

mattegan111 replied:
`notificationcontext` tells us where the notification came from, whether it be a post, or a comment on a post, or maybe the profile of the friend who sent us a friend request. Struggling to imagine how 'id' is related here...

nathanrobb replied:
> Struggling to imagine how 'id' is related here...

What id is going in here, post ids / comment ids?

--

I was thinking this field was the id of the notification in the db. It is a link to the object that caused the notification to be created. How are we distinguishing if this `notificationcontext` is from a `post` or from a `comment`?

mattegan111 replied:
The `notificationtype` will clarify what type of object caused the notification.

mattegan111 replied:
Doing it this way probably pushes alot of the work onto the frontend...

nathanrobb commented on server/models/Notification.js:
```diff
@@ -0,0 +1,31 @@
+
+	recipient: {
+		type: mongoose.Schema.Types.ObjectId,
+	},
```
Did we want to ref the user here?

nathanrobb commented on server/models/Notification.js:
```diff
@@ -0,0 +1,31 @@
+		required: true,
+	},
+
+	sender: {
+		type: mongoose.Schema.Types.ObjectId,
+	},
```
Is the user that triggered the notification to send? E.g. I liked your comment so I 'sent' a notification to you (the 'recipient')?

mattegan111 replied:
Yep!

nathanrobb replied:
Can add a `ref` here to the user too?

nathanrobb commented on server/models/Notification.js:
```diff
@@ -0,0 +1,31 @@
+		type: mongoose.Schema.Types.ObjectId,
+	},
+
+	read: {
+		type: Boolean,
+		required: true,
+	},
```
Default `false` or is that implied via mongo db/mongoose?

mattegan111 replied:
Let's add it.

nathanrobb commented on server/routes/notifications.js:
```diff
@@ -0,0 +1,120 @@
+router.get(
+	'/',
+	auth,
+	errorCatcher(async (req, res) => {
```
Did we want an optional query param to indicate just get my unread notifications? For a notification bubble later?

mattegan111 replied:
Makes sense, might come back to the idea.

nathanrobb commented on server/routes/notifications.js:
```diff
@@ -0,0 +1,120 @@
+			date: -1,
+		});
+
+		if (!notifications) {
+			return res
+				.status(404)
+				.json({ msg: 'No Notifications found' });
+		}
```
I feel like this might be okay as a 200 with an empty array as the response?

mattegan111 replied:
Yes! Not having any notifications is not an error on the users part.

nathanrobb commented on server/routes/notifications.js:
```diff
@@ -0,0 +1,120 @@
+	})
+);
+
+// @route   POST api/notifications
```
I feel like we'd largely be wanting to make new notifications from backend server side code? E.g. in the comment/like route handlers as side effects.

I can't think of when we'd want to create a new notification via an api.

nathanrobb reviewed (commented):
Nice, some good improvement comments here.

If you want to implement any lemme know and I'll approve later, just leaving some comments now.

mattegan111 committed:
Merge branch 'comment-model' into notification-model

nathanrobb commented on server/routes/notifications.js:
```diff
@@ -0,0 +1,120 @@
+		const notifications = await Notification.find({
+			receiver: req.user.id,
+		}).sort({
+			date: -1,
```
```suggestion
			date: 'desc',
```

Like we discussed in another PR👍

nathanrobb reviewed (approved):
LGTM. I feel like we might need to tweak the api query params when we actually begin to use these but we can do that when we know what we want from a front-end point of view.

What we have here is good as a base and allows for easy updates in the future when the react app starts talking with the api's.

mattegan111 committed:
Add reference to relevant model

mattegan111 committed:
Make notification false by default

mattegan111 committed:
Return empty array when user has no notifications

Don't need to return a 400 error as the user has not erred. We provide an empty array so the frontend can render the same way regardless of the number of notifications.

mattegan111 committed:
Improve readability of descending by date
